Миграция для создания платежного счета и способы взаимодействия со счетами

mrnght opened the pull request.

mrnght2 committed:
Создал миграцию для создания таблица платежных счетов, а так же сущность в ORM

mrnght2 committed:
BJS2-84165
Создал полный функционал открытия, получения счета, а так же его блокировки и закрытия

mrnght commented:
тут в первую очередь нужно сравнить созданную мной таблицу в миграции и сущность с заданием в тз

mrnght commented on src/main/java/faang/school/accountservice/service/AccountServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+
+        Long min = 100_000_000_000L;
+        Long max = 100_000_000_000_000_000L;
+        Long randomNum = min + (long)(new SecureRandom().nextDouble() * (max - min + 1));
```
генерация уникальных счетов выполняется в рамках отдельной задачи, тут сделал затычку

JekaCAP commented on src/main/resources/db/changelog/changeset/V002__account-service_payment-account.sql:
```diff
@@ -0,0 +1,16 @@
+    version INT NOT NULL
+);
+
+CREATE INDEX owner_id_index ON account (id);
```
Индекс на id это бессмысленно т.к id проиндексирован автоматически, лучше по типу владельца или по id владельца

JekaCAP commented on src/main/java/faang/school/accountservice/service/AccountServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public AccountViewDto blockAccount(Long id, AccountUpdateDto updateDto) {
```
Я бы добавил проверок на состояние перед закрытием. 

Нельзя закрыть уже закрытый счёт 
Нельзя блокировать закрытый счёт

JekaCAP commented on src/main/java/faang/school/accountservice/service/AccountServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+    private final AccountRepository repository;
+
+    @Override
+    public AccountViewDto openAccount(AccountCreateDto createDto) {
```
Можно открыть счёт с одинаковыми номерами?

JekaCAP commented on src/main/java/faang/school/accountservice/service/AccountServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+
+@Service
+@RequiredArgsConstructor
+public class AccountServiceImpl implements AccountService {
```
Методы сервиса я бы обернул в транзакции. 
Метод get @Transactional(readOnly = true) т.к хиберу не надо будет отслеживать изменения объектов

JekaCAP commented on src/main/java/faang/school/accountservice/entity/account/Account.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Version
+    @Column(name = "version")
+    private Integer version = 1;
```
"Когда запись платежного аккаунта изменяется, значение поля версии и времени создания также обновляются. Это позволяет системе отслеживать и контролировать версии записей и обнаруживать, если два или более процесса пытаются изменить одну и ту же запись одновременно. Если два процесса пытаются обновить запись с одинаковым значением версии, система может определить конфликт и предпринять соответствующие действия, например, откатить одно из изменений или предоставить уведомление о конфликте для разрешения. Для этого применить Optimistic Lock блокировку"

Отрывов из ТЗ ↑↑↑.  Тебе по тз, нужно реализовать оптимистичную блокировку, поле с версией уже есть 
https://habr.com/ru/articles/858714/

Вот хорошая статья на эту тему

mrnght replied:
Аннотация @Version уже по умолчанию реализует оптимистичную блокировку. Но я добавил логику доп. попыток для этого случая.

mrnght2 committed:
BJS2-84165
Добавил обработку оптимистичной блокировки

RinatBeybutov commented on src/main/java/faang/school/accountservice/controller/AccountController.java:
```diff
@@ -0,0 +1,54 @@
+@RequiredArgsConstructor
+@RequestMapping("/accounts")
+public class AccountController {
+    private final AccountServiceImpl service;
```
у нас принято по интерфейсу внедрять

RinatBeybutov commented on src/main/java/faang/school/accountservice/entity/account/Account.java:
```diff
@@ -0,0 +1,74 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Entity
```
@Data вместе с @Entity - снег в башка попадет

mrnght replied:
а как тестить тогда, все verify сразу отлетают

RinatBeybutov commented on src/main/java/faang/school/accountservice/repository/AccountRepository.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
для наследника JpaRepository эту аннотацию указывать необязательно

mrnght2 committed:
BJS2-84165
Написал тесты и объединил изменения статусов в один метод, тем самым добавив возможность разблокировки счета
